tokens/resource_service_principal_secret.go

```diff
@@ -2,125 +2,128 @@ package tokens
 			}
-			secrets, err = api.listServicePrincipalSecrets(spnID)
+
+			err = common.StructToData(*res, spnSecretSchema, d)
 			if err != nil {
 				return err
 			}
-			// ugly hack because rpc does not return ID of created secret
-			for _, v := range secrets.Secrets {
-				if len(idSeen) > 0 && idSeen[v.ID] {
-					continue
-				}
-				d.SetId(v.ID)
-			}
-			return d.Set("secret", secret.Secret)
+			d.Set("lifetime", lifetime)
+			d.Set("service_principal_id", spId)
+			d.SetId(res.Id)
+			return nil
 		},
 		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			if c.Config.AccountID == "" {
-				return errors.New("must have `account_id` on provider")
+			ac, err := c.AccountClient()
+			if err != nil {
+				return err
 			}
-			api := NewServicePrincipalSecretAPI(ctx, c)
-			spnID := d.Get("service_principal_id").(string)
-			secrets, err := api.listServicePrincipalSecrets(spnID)
+			spId := d.Get("service_principal_id").(string)
+			spIdNumeric, err := strconv.ParseInt(spId, 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to convert service principal ID to numeric: %w", err)
+			}
+			secrets, err := ac.ServicePrincipalSecrets.ListByServicePrincipalId(ctx, spIdNumeric)
 			if err != nil {
 				return err
 			}
 			for _, v := range secrets.Secrets {
-				if v.ID != d.Id() {
+				if v.Id != d.Id() {
 					continue
 				}
-				return d.Set("status", v.Status)
-			}
-			return &apierr.APIError{
-				ErrorCode:  "NOT_FOUND",
-				StatusCode: 404,
-				Message:    "client secret not found",
+				// check if the token is expired, although in practice API just doesn't return expired tokens
+				if v.ExpireTime != "" {
+					expireTime, err := time.Parse(time.RFC3339, v.ExpireTime)
+					if err != nil {
+						return fmt.Errorf("failed to parse expire time: %w", err)
+					}
+					if time.Now().After(expireTime) {
+						log.Printf("[INFO] service principal secret with id %s is expired, recreating it", d.Id())
+						d.SetId("")
+						return nil
+					}
+				}
+				// copy fields that aren't part of the result
+				secret := d.Get("secret").(string)
+				lifetime := d.Get("lifetime").(string)
+				common.StructToData(v, spnSecretSchema, d)
+				d.Set("secret", secret)
+				d.Set("lifetime", lifetime)
+				d.Set("service_principal_id", spId)
+				return nil
 			}
+			// recreate it if not found
+			log.Printf("[INFO] service principal secret with id %s not found, recreating it", d.Id())
+			d.SetId("")
+			return nil
 		},
 		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
-			if c.Config.AccountID == "" {
-				return errors.New("must have `account_id` on provider")
+			ac, err := c.AccountClient()
+			if err != nil {
+				return err
+			}
+			spId := d.Get("service_principal_id").(string)
+			spIdNumeric, err := strconv.ParseInt(spId, 10, 64)
+			if err != nil {
+				return fmt.Errorf("failed to convert service principal ID to numeric: %w", err)
```

Comment:
We have this error message appear thrice, would be good to abstract this out so we reduce redundancy.